# Housing Price Predictor


This notebook is based off of the Kaggle intro to machine learning course. The code from this course is altered to reflect my data. I do not use the data from the kaggle course. Instead, I have used the RapidAPI to generate 200 listings from the Minneapolis, MN. I then convert the JSON Format into a Pandas DataFrame and get to work on learning decision trees.

In [1]:
import pandas as pd
import numpy as np

import json
import requests

In [2]:
#Explore structure of the data
filename = 'readable_realty_data.json'
with open(filename) as f:
    readable_realty_data = json.load(f)

readable_realty_data = readable_realty_data['properties']
print(len(readable_realty_data))

200


In [3]:
#extracting the data
prop_ids, urls, cities, lines, postal_codes, counties, lats, lons, prices, baths_fulls, bedrooms, building_sizes, lot_sizes  =[],[],[],[], [], [], [], [],[],[], [], [], []


for data in readable_realty_data:
    prop_id = data['property_id']
    url = data['rdc_web_url']
    city = data['address']['city']
    line = data['address']['line']   
    postal_code = data['address']['postal_code']
    county = data['address']['county']
    lat = data['address']['lat']
    lon = data['address']['lon']
    price = data['price']
    baths_full = data['baths_full']
    bedroom = data['beds']
    building_size = data['building_size']['size']
    lot_size = data['lot_size']['size']
    
    
    
    prop_ids.append(prop_id)
    urls.append(url)
    cities.append(city)
    lines.append(line)
    postal_codes.append(postal_code)
    counties.append(county)
    lats.append(lat)
    lons.append(lon)
    prices.append(price)
    baths_fulls.append(baths_full)
    bedrooms.append(bedroom)
    building_sizes.append(building_size)
    lot_sizes.append(lot_size)

In [4]:
# converting json dataset from dictionary to dataframe
df = pd.DataFrame([prop_ids,urls,cities,lines, postal_codes, counties, lats, lons, prices,baths_fulls, bedrooms, building_sizes, lot_sizes]).T



In [5]:
#Rename the data
df = df.rename(columns = {0:'prop_ids',1:'urls',2:'cities',3:'lines', 4:'postal_codes', 5:'counties', 6:'lats', 7:'lons', 8:'prices', 9:'baths_fulls',10:'bedrooms',11:'building_sizes', 12:'lot_sizes'})


In [6]:
df.head()

prop_ids                                               urls  \
0  M7927081518  https://www.realtor.com/realestateandhomes-det...   
1  M8412373846  https://www.realtor.com/realestateandhomes-det...   
2  M7324092858  https://www.realtor.com/realestateandhomes-det...   
3  M8405448159  https://www.realtor.com/realestateandhomes-det...   
4  M8627783588  https://www.realtor.com/realestateandhomes-det...   

             cities              lines postal_codes  counties     lats  \
0        Saint Paul    79 Cretin Ave N        55104    Ramsey   44.944   
1       Minneapolis   1716 Tyler St NE        55413  Hennepin  45.0055   
2             Eagan     2094 Quartz Ln        55122    Dakota  44.8001   
3       Robbinsdale  3628 Regent Ave N        55422  Hennepin  45.0214   
4  Saint Louis Park   2732 Joppa Ave S        55416  Hennepin  44.9531   

      lons  prices baths_fulls bedrooms building_sizes lot_sizes  
0 -93.1928  600000           2        4           3143     14375  
1 -93.2465  475000           3        5           2448      8276  
2 -93.2149  324900           2        4           1620      9017  
3 -93.3437  314999           2        4           2255      5227  
4 -93.3345  649900           3        3           2571      6098

In [7]:
df.describe()

prop_ids                                               urls  \
count           200                                                200   
unique          200                                                200   
top     M8284753737  https://www.realtor.com/realestateandhomes-det...   
freq              1                                                  1   

             cities            lines postal_codes  counties       lats  \
count           200              200          200       200  200.00000   
unique           37              200           62         7  200.00000   
top     Minneapolis  3431 Elliot Ave        55423  Hennepin   44.91405   
freq             57                1           13       131    1.00000   

            lons  prices  baths_fulls  bedrooms  building_sizes  lot_sizes  
count   200.0000     200          200       200             200        200  
unique  199.0000     105            4         7             186         95  
top     -93.3725  350000            2         3            2400       5227  
freq      2.0000       9          133        93               3         17

# Decision Trees


I begin with a simple Decision tree and work my way up.

In [52]:
#defining dependent variable
y = df.prices

In [53]:
df_features = ['lats','lons','postal_codes','baths_fulls','bedrooms','building_sizes','lot_sizes']


In [54]:
#defining Indepenent variables
X = df[df_features]

In [55]:
X.describe()

lats      lons postal_codes  baths_fulls  bedrooms  \
count   200.00000  200.0000          200          200       200   
unique  200.00000  199.0000           62            4         7   
top      44.91405  -93.3725        55423            2         3   
freq      1.00000    2.0000           13          133        93   

        building_sizes  lot_sizes  
count              200        200  
unique             186         95  
top               2400       5227  
freq                 3         17

In [12]:
X.head()

lats     lons postal_codes baths_fulls bedrooms building_sizes lot_sizes
0   44.944 -93.1928        55104           2        4           3143     14375
1  45.0055 -93.2465        55413           3        5           2448      8276
2  44.8001 -93.2149        55122           2        4           1620      9017
3  45.0214 -93.3437        55422           2        4           2255      5227
4  44.9531 -93.3345        55416           3        3           2571      6098

In [13]:
from sklearn.tree import DecisionTreeRegressor

In [14]:
Min_model = DecisionTreeRegressor(random_state=1)

In [15]:
Min_model.fit(X,y)

DecisionTreeRegressor(random_state=1)

In [16]:
print("Making predictions for the following 5 houses:")
print(X.head())
print("The predictions are")
print(Min_model.predict(X.head()))


Making predictions for the following 5 houses:
      lats     lons postal_codes baths_fulls bedrooms building_sizes lot_sizes
0   44.944 -93.1928        55104           2        4           3143     14375
1  45.0055 -93.2465        55413           3        5           2448      8276
2  44.8001 -93.2149        55122           2        4           1620      9017
3  45.0214 -93.3437        55422           2        4           2255      5227
4  44.9531 -93.3345        55416           3        3           2571      6098
The predictions are
[600000. 475000. 324900. 314999. 649900.]


In [17]:
#model validation

# Mean Absolute Error
This is taking the mean of the error (actual−predicted) and putting it in absolute value terms. The lower this value the more accurate the model predicted the prices.

In [18]:
df.head()

prop_ids                                               urls  \
0  M7927081518  https://www.realtor.com/realestateandhomes-det...   
1  M8412373846  https://www.realtor.com/realestateandhomes-det...   
2  M7324092858  https://www.realtor.com/realestateandhomes-det...   
3  M8405448159  https://www.realtor.com/realestateandhomes-det...   
4  M8627783588  https://www.realtor.com/realestateandhomes-det...   

             cities              lines postal_codes  counties     lats  \
0        Saint Paul    79 Cretin Ave N        55104    Ramsey   44.944   
1       Minneapolis   1716 Tyler St NE        55413  Hennepin  45.0055   
2             Eagan     2094 Quartz Ln        55122    Dakota  44.8001   
3       Robbinsdale  3628 Regent Ave N        55422  Hennepin  45.0214   
4  Saint Louis Park   2732 Joppa Ave S        55416  Hennepin  44.9531   

      lons  prices baths_fulls bedrooms building_sizes lot_sizes  
0 -93.1928  600000           2        4           3143     14375  
1 -93.2465  475000           3        5           2448      8276  
2 -93.2149  324900           2        4           1620      9017  
3 -93.3437  314999           2        4           2255      5227  
4 -93.3345  649900           3        3           2571      6098

In [56]:
# Filter rows with missing price values
filtered_min_data = df.dropna(axis=0)
# Choose target and features
y = filtered_min_data.prices
min_features = ['lats','lons','postal_codes','baths_fulls','bedrooms','building_sizes','lot_sizes']
X = filtered_min_data[min_features]

from sklearn.tree import DecisionTreeRegressor
# Define model
min_model = DecisionTreeRegressor()
# Fit model
min_model.fit(X, y)

DecisionTreeRegressor()

In [20]:
from sklearn.metrics import mean_absolute_error

predicted_home_prices = Min_model.predict(X)
mean_absolute_error(y, predicted_home_prices)

0.0

The value above seems to be way overfitted, meaning it does not generalize the results. Using a validation technique of train, test, split we can hopefully build an accurate model that generalizes the results better. The train, test, split approach "chops up" the obversations between two groups, a test and training group. We'll run the training group to precict price values and compare them to the test set of values.

In [57]:
from sklearn.model_selection import train_test_split

# Splitting the y and X variables. Val_y is the test group.
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 1)
# Defining model
min_model = DecisionTreeRegressor(random_state = 1)
# Fitting model
min_model.fit(train_X, train_y)

# get predicted prices on validation data
val_predictions = min_model.predict(val_X)
print(mean_absolute_error(val_y, val_predictions))

150870.0


The value driven through the train,test,split approach equals $150,870.0. This is a drastic increase indicating that the past test was possibly too accurate

How can we know when we are going to overfit or underfit the data before we start?
Kaggle recommends the max_leaf_nodes argument.
This function passes five arguments into the objectives below. It then calculates the model which uses the decision tree finding many MAEs for the many different number of nodes within the decision tree model. It then fits the data, predicts the values, and outputs the MAEs

In [30]:
from sklearn.metrics import mean_absolute_error
from sklearn.tree import DecisionTreeRegressor

def get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y):
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes, random_state=1)
    model.fit(train_X, train_y)
    preds_val = model.predict(val_X)
    mae = mean_absolute_error(val_y, preds_val)
    return(mae)

In [40]:
# compare MAE with differing values of max_leaf_nodes
for max_leaf_nodes in [5, 50, 500, 5000]:
    my_mae = get_mae(max_leaf_nodes, train_X, val_X, train_y, val_y)
    print("Max leaf nodes: %d  \t\t Mean Absolute Error:  %d" %(max_leaf_nodes, my_mae))

Max leaf nodes: 5  		 Mean Absolute Error:  178160
Max leaf nodes: 50  		 Mean Absolute Error:  151289
Max leaf nodes: 500  		 Mean Absolute Error:  151670
Max leaf nodes: 5000  		 Mean Absolute Error:  151670


In [39]:
candidate_max_leaf_nodes = [5, 25, 50, 100, 250, 500]
# Write loop to find the ideal tree size from candidate_max_leaf_nodes
scores = {leaf_size: get_mae(leaf_size, train_X, val_X, train_y, val_y) for leaf_size in candidate_max_leaf_nodes}
print("Max leaf nodes: %d  \t\t Mean Absolute Error:  %d" %(max_leaf_nodes, my_mae))

Max leaf nodes: 5000  		 Mean Absolute Error:  151670


In [37]:
best_tree_size = min(scores, key=scores.get)

We can see that as we increase nodes the MAE falls then rises. Everything over 500 seems to be just as good as 500. 50 is given as the best model because it has the lowest MAE. This is because as the nodes increase the model is shrinking and predicting prices on a smaller sample. The opposite is true for a lower node and is over-generalizing the model.


lets re-run the code with the data we know now.

In [50]:
final_model = DecisionTreeRegressor(max_leaf_nodes = best_tree_size, random_state=1)

# fit the final model and uncomment the next two lines
final_model.fit(train_X, train_y)
final_model.fit(train_X, train_y)
predicted_home_prices = final_model.predict(train_X)
print(final_model.predict(X.head()))

[630000.         466880.         347786.48648649 414048.16129032
 466880.        ]


In [51]:
# get predicted prices on validation data
val_predictions = final_model.predict(val_X)
print(mean_absolute_error(val_y, val_predictions))

147385.8504509348


We can see that the MAE has fallen using the Max leaf nodes test. This should be expected because we optimizing our model given what would have produced the best MAE. It's overall interesting to see how far the value fell compared to the train X and y of the original decision tree. The first test value: 150,870.0, the second test 147,385.8504. Roughly a 2-3k decline in MAE.


# Random Forest


Random forest essentially uses many trees and averages the outcomes of each tree.

In [49]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

forest_model = RandomForestRegressor(random_state=1)
forest_model.fit(train_X, train_y)
min_preds = forest_model.predict(val_X)
print(mean_absolute_error(val_y, min_preds))

111751.546


This value is a huge leap in the accuracy of the model compared to just one decision tree. For reference:
The first test value: 150,870.0
the second test: 147,385.8504
third test (Random Forest): 111,751.546